### DISCUSSION TITLE
[General] Fair pubdata price too high prevents tx inclusion

### DISCUSSION THREAD

**Discussion by nambrot** [zkSync-Community-Hub/zksync-developers#825] [General] [answered]
### Team or Project

Hyperlane / Treasure

### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

Trying to submit transaciton but getting this error: `Fair pubdata price too high`

**Comment by alhenaai** [comment]
It seems like you are encountering an issue with the pubdata price being too high, which prevents the inclusion of your transaction on the ZKsync network. This issue might be related to the mechanism by which ZKsync Era charges for pubdata. Here's a brief overview of how it works:

1. **Base Pubdata Spent**: The amount of pubdata already published is fixed.
2. **Validation and Execution**: The transaction is validated and executed. If the transaction does not cover enough funds for itself, it should be rejected.
3. **Gas and Pubdata Costs**: The system checks if the gas provided covers the pubdata costs. If not, the transaction is reverted, but the user still pays the fee for the computation.

For more detailed information on how pubdata fees are calculated and the implications for transaction execution in ZKsync Era, you can refer to the [ZKsync documentation on pubdata charges](https://docs.zksync.io/zksync-protocol/rollup/fee-model/how-we-charge-for-pubdata).

If you continue to face issues, I recommend discussing this further with the development community for more specific guidance. You can join the discussions on [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by nambrot** [comment]
It turns out that the gasPrice that was set was not high enough, but our service is just getting it from the RPC via `eth_feeHistory ` in the 1559 fashion and then applying ethers-rs standard estimation with that

**Comment by shahar4** [comment]
@nambrot happy this is solved and thanks for the update you solved it. Can you provide some more details on where you saw this error though?

**Reply by daniel-savu** [reply]
from the RPC node and the error was propagated via ethers-rs:
```
ContractError(MiddlewareError(MiddlewareError(MiddlewareError(MiddlewareError(MiddlewareError(JsonRpcClientError(JsonRpcError(JsonRpcError { code: 3, message: "failed to validate the transaction. reason: Validation revert: virtual machine entered unexpected state. Please contact developers and provide transaction details that caused this error. Error description: Assertion error: Fair pubdata price too high", data: Some(String("0x")) }))))))))
```

**Reply by shahar4** [reply]
Thanks! What action were you doing to cause this?

**Reply by daniel-savu** [reply]
it was an `eth_sendRawTransaction` rpc call to `PrometheusJsonRpcClient(Http { id: 213652, url: https://rpc.treasure.lol/ })`

**Reply by daniel-savu** [reply]
with the gas set to `max_priority_fee_per_gas: Some(100000000), max_fee_per_gas: Some(719033514608)`

**Reply by daniel-savu** [reply]
here's our eip-1559 gas estimation logic btw, wonder if you're using anything different: https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/bb44f9b51431e0ec19b62195a74bbf7fc0693bc7/rust/main/chains/hyperlane-ethereum/src/tx.rs#L205

**Reply by shahar4** [reply]
@daniel-savu thanks for the context. Let me get back to you. If in the meanwhile you’re experiencing any other issues please reach out again

**Reply by daniel-savu** [reply]
@shahar4 I wouldn't say the issue is fully fixed, even if Nam marked it as "resolved". Currently we just 2x the gas price estimate we get from the function I linked above, but would like to understand why the node outright rejects any transactions with an insufficient gas price instead of accepting them and keeping them in the mempool

**Reply by nambrot** [reply]
100% would say that this is not resolved. This behavior differs from other environments and is critical for developers to build production experiences on top of zksync stack chains

**Reply by nambrot** [reply]
@sarahschwartz can you unmark it as the answer since it's not actually resolved

**Reply by karelvuong** [reply]
@shahar4 flagging for you

**Reply by dimazhornyk** [reply]
hi guys! it would be great if you could share the exact calldata and addresses to help us debug the underlying issue

**Reply by shahar4** [reply]
Hey guys! Yes we’re on it and trying to get to the bottom of this here. Agreed on unresolving.

**Reply by tkporter** [reply]
Here's an example transaction pulled from our logs:
```
MonitoredTransaction { hash: 0x7f5fa39c6fce7ebf8199fd7eb230572d8524d29b56a98b55478bce4db6547419, inner: Eip1559(Eip1559TransactionRequest { from: Some(0x74cae0ecc47b02ed9b9d32e000fd70b9417970c5), to: Some(0x2e29fe39496a56856d8698bd43e1df4d0ce6266a), gas: Some(1557246), value: None, data: Some(Bytes(0x82ad56cb000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000034000000000000000000000000000000000000000000000000000000000000006200000000000000000000000009bbdf86b272d224323136e15594fdce487f40ce70000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000002447c39d130000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000001800000000000000000000000000000000000000000000000000000000000000107000000000000000000000000748040afb89b8fdbb992799808215419d36a09306c55c7886663994a6ab5fe746ab5a60d2868022f233c908e46aebfab8241c6d800173ad37369420c636cad6a8ec42ca6bbb8bc5963ce1c5c609c4493dbe5b677562c003300a3ad5f38bbac07a45d784ca30fe9b1126bf63c74bd7134fe2603ccd8fdbfea1cc0578edba4d6991904f44942f29b4ca1291aa2f17507ce43981c6f64056bb2b968217d7839af38b7075fed0ce97b5c4532c4d04096e726f1d5c634e1f58bfc5a1c16ac837a3ba6400782942e0e4b7940507a20f5c499a4565591ae6e78b48c36442875e5cea10d32b97888bd67e4bc8668fde13011d6c5abf7be6c8a651b78866e1c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008d0300173ae80000a4b1000000000000000000000000240d04a70b038369c8df703b78b3b47332eee1160000eeee00000000000000000000000001c94f24f8d72bb9c3f61c4ed0b9b86bfc23badd0000000000000000000000007e29a2aa8cc79908e00b9ef6970a52bdec942f6000000000000000000000000000000000000000000000000053444835ec58000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000009bbdf86b272d224323136e15594fdce487f40ce70000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000002447c39d130000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000001800000000000000000000000000000000000000000000000000000000000000107000000000000000000000000748040afb89b8fdbb992799808215419d36a0930c2ea44399594943d4e789dd60572d2d241bcf3843b2cbce33f7b4953f07e06b500173ad5bfb84aba1ae4e721dd267ab256bac249ccd067d9126e5ef6929f070d0f4361e173c6c3930887bc446306cc3519bd2ed2fe3116b8575b4bc50b6a62f1221e858f1c6d845d8eccb825b1f3d7063c95ffefbba9aa7381e573910f89cd07170777bee97c24c18f2fce4eeeaa0debbe149952449d464a32e0255515707437b9dcfef0ee1c46efc0aef076bf88762063c1832988549dd968d22f03330909b7595b1f447ef0454cf495cf577b53cc6beeb4dd2fa6ad6489233359cde19730dbc64708c2ddf01c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008d0300173aea0000a4b1000000000000000000000000240d04a70b038369c8df703b78b3b47332eee1160000eeee00000000000000000000000001c94f24f8d72bb9c3f61c4ed0b9b86bfc23badd000000000000000000000000253092c195c09fa8638ca4df56601dc8ae738c51000000000000000000000000000000000000000000000000139f89822067000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000009bbdf86b272d224323136e15594fdce487f40ce70000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000002447c39d130000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000001800000000000000000000000000000000000000000000000000000000000000107000000000000000000000000748040afb89b8fdbb992799808215419d36a093034323d260b417e5fd6357aaacc79f23ec443fa8f00ae78896916ce4efdbe721b00173ad11740db32294ca2fd353ded6354717113531763810ad6726a00ff7aa5fd78b9e855e8209029e3fc366fcdc32a0541a20a15b9bfc64f50497cf1eab845a3ad37a71ce2e22d6aca9081b4f8ab6c16f4d1f49290157fef230ffd58ca5d653ba97638030ef397bbcf6b79185358dd1c34dd90fb9bef0ca2534c7279cded1d640f5667351b9a5894de0a210ae1065b71174af294ecc921f8f29a857af4e048a4aab4b900fe64872d85f4c92c2a2f3640e4d72627baa2516229aa1f5da9127e47796962a80d1c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008d0300173ae60000a4b1000000000000000000000000240d04a70b038369c8df703b78b3b47332eee1160000eeee00000000000000000000000001c94f24f8d72bb9c3f61c4ed0b9b86bfc23badd000000000000000000000000d846088e0d4f97ed5e17b24d0ded2c640e4f313c0000000000000000000000000000000000000000000000008ac7230489e800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000)), nonce: Some(5957), access_list: AccessList([]), max_priority_fee_per_gas: Some(455000000), max_fee_per_gas: Some(3378550271912), chain_id: Some(61166) }), creation_time: Instant { tv_sec: 23961, tv_nsec: 753381486 }, block: None }
```

The chain is Treasure, but this is using the zkSync stack - some metadata on that chain if you're unfamiliar https://github.com/hyperlane-xyz/hyperlane-registry/blob/main/chains/treasure/metadata.yaml

**Reply by tkporter** [reply]
Would we have more success using https://docs.zksync.io/zksync-protocol/api/zks-rpc#zks_estimatefee instead of typical eth APIs for figuring out gas prices to use?

**Reply by JackHamer09** [reply, marked as answer]
@tkporter Yes, it is recommended to use `zks_estimateFee` for estimating transaction fees on ZKsync. This should work out of the box with `zksync-ethers` and viem SDKs.
Please, let me know if this helps!

**Reply by dimazhornyk** [reply]
@tkporter the tx you shared has executed [successfuly](https://treasurescan.io/tx/0x7f5fa39c6fce7ebf8199fd7eb230572d8524d29b56a98b55478bce4db6547419), any chance you have the input of the failed ones, that were rejected by the server?
